src/client/cl_main.cpp

```diff
@@ -4043,3 +4046,128 @@ void CL_GetVMGLConfig(vmglconfig_t *vmglconfig) {
 	vmglconfig->stereoEnabled = cls.glconfig.stereoEnabled;
 	vmglconfig->smpActive = cls.glconfig.smpActive;
 }
+
+void CL_ShaderStateChanged( void ) {
+	// Originally copied from CG_ShaderStateChanged, but rewritten to avoid issues of the original implementation and to
+	// support settings
+	char originalShader[MAX_QPATH];
+	char newShader[MAX_QPATH];
+	char settings[32];
+	char *timeOffset;
+	char *lightmapMode;
+	char *styleMode;
+
+	shaderRemapLightmapType_t lightmapModeValue;
+	shaderRemapStyleType_t styleModeValue;
+
+	const char *curPos = cl.gameState.stringData + cl.gameState.stringOffsets[ CS_MV_REMAPS ];
+	const char *endPos = curPos + strlen( curPos );
+
+	// Check configstring for prefix to avoid misinterpreting configstrings from mods
+	if ( !curPos || !*curPos || strncmp(curPos, "mvremap:", 8) ) return;
+
+	// Handling of these is really ugly. I originally wanted to handle them like the base game/cgame modules do, but
+	// those are prone to injections. Summary of delimeter issues:
+	//  - base uses '=', ':' an '@', all of them can be used in shader names
+	//  - .shader files don't seem to support spaces in shader names, but texture paths support spaces
+	//  - .shader files allow most other characters, including backspace '\'
+	//  -> using a delimeter is not a reliable option
+
+	// The next best thing seemed to be to give the length of the shader before its name to allow all characters to be
+	// used. As I don't want to waste any space (configstrings containing shader paths are wasteful as it is), I decided
+	// to encode the length in a single byte (MAX_QPATH is 80, so a single byte can easily hold the length). However the
+	// netcode and the handling of quotes lead to other issues:
+	//  - 34 maps to 32, thus 32 could mean either of these numbers while 34 doesn't exist
+	//  - 37 and >127 map to 46, thus 37 doesn't exist and 46 could mean 37, 46 or >127
+	//  -> the biggest amount of consecutive numbers without multiple meanings spans from 47 to 127; as MAX_QPATH is 80
+	//     this should fix exactly
+	//  -> could also use the range from 38 to 127, because - when isolating the range - there are no duplicates within
+	//     it; this would allow for a size of up to 89, which is unlikely to be useful if MAX_QPATH ever gets increased
+	//  -> using 47 to 127
+
+	// Skip prefix
+	curPos += 8;
+
+	int length;
+
+	// Parse configstring
+	while ( curPos && *curPos )
+	{
+		// Read original shader
+		if ( curPos >= endPos ) break;
```

Comment:
How about some error reporting for easier debugging with DPrintf? maybe a little goto instead of break would come handy here? Would be nice to also print the broken piece of remap string because these get pretty long.

Reply:
I don't think we need that many details. If the remap string is broken it's probably something wrong with the configstring the server built and that could be broken in so many ways it's hard to pinpoint the exact issue. Something may have been offset early on and we only run into the issue later in the string. A DPrintf about incorrectly encoded length for source/destination/settings should be enough info in my opinion.

Reply:
as I wrote - asking for this only if it's simple enough. I have bad experience debugging parsers that give 0 info about what's broken, and work on large input :) how about just adding "%d" curPos - startPos to DPrintf? just to give some general hint.

Reply:
ok I see why it's unlikely to give any useful info when remap string is broken - it will probably parse a lot of gibberish and then break much later. that's too bad but if we go with this format - fine.

Another thing - when q_strncpyz is called with destsize == 0, it throws ERR_FATAL. I don't think we want server to be able to crash jk2mv clients, at least not in features we add. so let's make it if (length < 1)